### PR TITLE
[WIP] Add scmRepo to pipeline

### PR DESCRIPTION
**Pull request by catto**
I added a `scmRepo` to a modelConfig for a pipeline. screwdriver-cd/screwdriver#160
I'm not sure this is correct but I think I shouldn't call getRepoId at [here](https://github.com/screwdriver-cd/screwdriver/pull/204/files#diff-1eb6375f360812e82460d42f926123b1L77).

### CODE REVIEW

**Comment by stjohnjohnson** [comment]
Why not use the new function added here: https://github.com/screwdriver-cd/models/pull/106/files

**Comment by catto** [comment]
Ahh.. I completely misunderstood.
I'll change these to use `refreshScmRepo` and add duplicate checking instead of it.

**Comment by catto** [comment]
Hmm.. we can call `refreshScmRepo` if a pipeline already exists. But `id` for new pipeline will be generated using `scmRepo.id` before a creation of new pipeline, so we can't call `refreshScmRepo`. We need to get scmRepo and call `super.create` with scmRepo, don't we?

**Comment by stjohnjohnson** [comment]
@catto where can we move that function then?  Maybe make it static somehow?

**Comment by catto** [comment]
@stjohnjohnson 🙆 I'll fix them.
I was just wondering if you had design discussions and how much code I can change.

**Comment by catto** [comment]
I made it static and added duplication checks. If they are okay, I will change pipeline plugin to catch the error and return a conflict error.

File: lib/pipeline.js [hunk 1]
```diff
@@ -123,6 +115,35 @@ class PipelineModel extends BaseModel {
     }
 
     /**
+     * Get the first admin username of the pipeline.
+     * @method getFirstAdminName
+     * @return {String}
+     */
+    // TODO: Get the first admin. Validate if it's valid. If not, remove from admins field,
+    //       and retry with the next admin.
+    getFirstAdminName() {
```
Comment (petey) [inline lib/pipeline.js]: `getFirstAdminName`, and `getAdmin` should appear in source order before they are used (in `get admin`)

File: lib/pipeline.js [hunk 2]
```diff
@@ -287,20 +308,31 @@ class PipelineModel extends BaseModel {
...
+    }
+
+    /**
+     * Get the SCM Repo based on the SCM URL
+     * @method getScmRepo
+     * @param  {String}  username The admin username for the repository of the scmUrl
+     * @param  {String}  scmUrl   The scmUrl for the SCM Repo object
+     * @return {Promise}
+     */
+    static getScmRepo(username, scm, scmUrl) {
```
Comment (petey) [inline lib/pipeline.js]: `getScmRepo` should appear in source order before it is used

File: lib/pipelineFactory.js [hunk 3]
```diff
@@ -42,8 +42,22 @@ class PipelineFactory extends BaseFactory {
         const modelConfig = hoek.applyToDefaults({
             createTime: (new Date()).toISOString()
         }, config);
+        const admin = config.admins[0];
 
-        return super.create(modelConfig);
+        return Pipeline.getScmRepo(admin, this.scmPlugin, config.scmUrl)
```
Comment (petey) [inline lib/pipelineFactory.js]: I don't really understand why this is something the pipeline model has to be a part of, other than it is a convenient place to have static methods to get a user model. 

On the other hand, the pipelineFactory could have a method that fetches a user model, and scmPlugin could have a method `getRepoData(userModel, scmUrl)` or whatever that does the lookup.

**Review by petey — CHANGES_REQUESTED** [review]

File: lib/pipelineFactory.js (hunk 3 above)
Comment (catto) [reply lib/pipelineFactory.js]: That's reasonable. I have moved `getScmRepo` to the pipelineFactory. I think `getRepoData` in the scmPlugin is not necessary in this case.

**Comment by petey** [comment]
Closing in favor of work in #120

**Pull request closed without merging**